gcp_variant_transforms/transforms/infer_undefined_headers.py

```diff
@@ -0,0 +1,119 @@
+from gcp_variant_transforms.transforms import merge_headers
+
+
+__all__ = ['InferUndefinedHeaderFields']
+
+
+_FLAG = 'Flag'
+_STRING = 'String'
+
+
+class _InferUndefinedHeaderFields(beam.DoFn):
+  """Extracts undefined header fields from ``Variant`` record."""
+
+  def _get_field_count(self):
+    # TODO(nmousavi): return filed count by conditioning on field_value type
+    # (list, bool, o/w) once variant_to_bigquery code can deduct type for
+    # missing header fields from extracted final headers. Currently it inserts
+    # fields with missing header to bigquery as a repeated field.
+    return field_counts['.']
+
+  def _get_field_type(self, field_value):
+    if isinstance(field_value, bool):
+      return _FLAG
+    else:
+      return _STRING
+
+  def _infer_undefined_info_fields(self, variant, defined_headers):
+    infos = {}
+    for info_field_key, variant_info in variant.info.iteritems():
+      info_field_value = variant_info.data
+      if not defined_headers or info_field_key not in defined_headers.infos:
+        if info_field_key in infos:
+          raise ValueError(
+              'Invalid VCF file. Duplicate INFO field in variant {}'.format(
+                  variant))
+        infos[info_field_key] = Info(info_field_key,
+                                     self._get_field_count(),
+                                     self._get_field_type(info_field_value),
+                                     '',  # NO_DESCRIPTION
+                                     '',  # UNKNOWN_SOURCE
+                                     '')   # UNKNOWN_VERSION
+    return infos
+
+  def _infer_undefined_format_fields(self, variant, defined_headers):
+    formats = {}
+    for call in variant.calls:
+      for format_key, format_value in call.info.iteritems():
+        if not defined_headers or format_key not in defined_headers.formats:
+          if format_key in formats:
+            raise ValueError(
+                'Invalid VCF file. Duplicate FORMAT field in variant {}'.format(
+                    variant))
+          formats[format_key] = Format(format_key,
+                                       self._get_field_count(),
+                                       self._get_field_type(format_value),
+                                       '') # NO_DESCRIPTION
+      # No point in proceeding. All other calls have the same FORMAT.
+      break
+    return formats
+
+  def process(self, variant, defined_headers):
+    infos = self._infer_undefined_info_fields(variant, defined_headers)
+    formats = self._infer_undefined_format_fields(variant, defined_headers)
+    if infos or formats:
+      yield vcf_header_io.VcfHeader(infos=infos, formats=formats)
+
+class InferUndefinedHeaderFields(beam.PTransform):
+  """Extracts undefined header fields from ``Variant`` records."""
+
+  def __init__(self, defined_headers):
+    """Initializes the transform.
+    Args:
+      defined_headers (:class:``vcf_header_io.VCFHeader``): side input
```

Comment:
nit: when using `:class:` , you only need one backquote.

Reply:
Done